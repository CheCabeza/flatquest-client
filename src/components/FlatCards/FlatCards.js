import * as React from 'react';
import {View, Text, Image, Pressable, TouchableOpacity} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import Icon2 from 'react-native-vector-icons/MaterialCommunityIcons';
import Icon3 from 'react-native-vector-icons/FontAwesome5';
import {connect} from 'react-redux';
import {PropTypes} from 'prop-types';
import {addFavorite} from './../../redux/actions/actionCreators';
import {useNavigation} from '@react-navigation/native';
import styles from '../List/styles';

function FlatCards({flats, dispatch, user, flatStore}) {
  const navigation = useNavigation();
  function addFav(flatId) {
    dispatch(
      addFavorite({
        id: user.user._id,
        favorites: [...user.user.favorites, flatId],
        bearerToken: user.token,
      }),
    );
  }

  function removeFav(flatId) {
    let arr = [...user.user.favorites];
    arr.splice(arr.indexOf(flatId), 1);
    dispatch(
      addFavorite({
        id: user.user._id,
        favorites: arr,
        bearerToken: user.token,
      }),
    );
  }

  function findIcon(flatId) {
    const flat = flatStore.find(element => element._id === flatId).face;
    const highestRate = flat.sort(function (a, b) {
      return parseFloat(b.votes) - parseFloat(a.votes);
    });
    return {type: highestRate[0].type, votes: highestRate[0].votes};
  }
  return flatStore.map((flatsData, index) => (
    <Pressable
      testID="FlatCard"
      key={index}
      onPress={() => navigation.navigate('Details', {flatId: flatsData._id})}>
      <View style={styles.box} key={flatsData.index}>
        <Image
          style={styles.img}
          source={{
            uri: `${flatsData.photos[0]}`,
          }}
        />
        <View style={styles.info}>
          <Text>{flatsData.address}</Text>
          <Text>{flatsData.neighborhood}</Text>
          <Text>{flatsData.price} â‚¬</Text>
          <View style={styles.iconSection}>
            <Text>{flatsData.toilet} </Text>
            <Icon style={styles.icon} name="bath" color="#000" size={25} />
            <Text>{flatsData.rooms} </Text>
            <Icon style={styles.icon} name="bed" color="#000" size={25} />
            <Text>{flatsData.m2}m2</Text>
            <Icon2
              style={styles.icon}
              name="tape-measure"
              color="#000"
              size={25}
            />
            <View style={[styles.icon_container, {top: -85, right: -30}]}>
              <TouchableOpacity
                testID="FlatCardsButton"
                onPress={() => {
                  user.user.favorites.includes(flatsData._id)
                    ? removeFav(flatsData._id)
                    : addFav(flatsData._id);
                }}>
                {user.user.favorites.includes(flatsData._id) ? (
                  <Icon name="heart" color="#EE3BC5" size={25} />
                ) : (
                  <Icon name="heart" color="#000" size={25} />
                )}
              </TouchableOpacity>
            </View>

            <View style={[styles.icon_container, {top: -260, right: -10}]}>
              <Icon3
                name={findIcon(flatsData._id).type}
                color="#000"
                size={35}
              />
              <View style={styles.miniIcon_container}>
                <Text>{findIcon(flatsData._id).votes}</Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </Pressable>
  ));
}

FlatCards.propTypes = {
  flats: PropTypes.array.isRequired,
  dispatch: PropTypes.func.isRequired,
  filteredFlats: PropTypes.array.isRequired,
};

function mapStateToProps(store) {
  return {
    flats: store.flats,
    filteredFlats: store.filteredFlats,
    user: store.user,
  };
}

export default connect(mapStateToProps)(FlatCards);
