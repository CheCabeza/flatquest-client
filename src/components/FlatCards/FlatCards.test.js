import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {useNavigation} from '@react-navigation/native';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import FlatCards from './FlatCards';
import thunk from 'redux-thunk';
jest.mock('@react-navigation/native');

describe('Given a FlatCards component', () => {
  const mockStore = configureStore([thunk]);
  let store;
  let navigation;
  let flats = [
    {
      _id: 476,
      photos: 'a',
      pros: ['glass', 'folder-open'],
      cons: ['money', 'copy'],
      reviews: ['Mierda de test'],
      face: [
        {type: 'frown', votes: '7'},
        {type: 'grin', votes: '3'},
        {type: 'smile', votes: '1'},
        {type: 'meh', votes: '0'},
      ],
    },
  ];

  beforeEach(() => {
    store = mockStore({
      flats: [
        {
          _id: 123,
          photos: 'a',
          pros: ['glass', 'folder-open'],
          cons: ['money', 'copy'],
          reviews: ['Mierda de test'],
          face: [
            {type: 'frown', votes: '7'},
            {type: 'grin', votes: '3'},
            {type: 'smile', votes: '1'},
            {type: 'meh', votes: '0'},
          ],
        },
      ],
      filteredFlats: [
        {
          _id: 123,
          photos: 'a',
          pros: ['glass', 'folder-open'],
          cons: ['money', 'copy'],
          reviews: ['Mierda de test'],
          face: [
            {type: 'frown', votes: '7'},
            {type: 'grin', votes: '3'},
            {type: 'smile', votes: '1'},
            {type: 'meh', votes: '0'},
          ],
        },
      ],
      user: {
        user: {
          email: 'ho@la.com',
          pass: '123123123',
          favorites: [123, 456],
        },
      },
    });

    useNavigation.mockReturnValueOnce({
      navigate: jest.fn(),
    });
    navigation = {
      navigate: jest.fn(),
    };
  });

  test('Should render the FlatCards of page', () => {
    const Screen = render(
      <Provider store={store}>
        <FlatCards flatStore={flats} navigation={navigation} />
      </Provider>,
    );
    expect(Screen).toMatchSnapshot();
  });

  test('When flats_id !== user.user.favorites addFav is invoked', () => {
    flats = [
      {
        _id: 476,
        photos: 'a',
        pros: ['glass', 'folder-open'],
        cons: ['money', 'copy'],
        reviews: ['Mierda de test'],
        face: [
          {type: 'frown', votes: '7'},
          {type: 'grin', votes: '3'},
          {type: 'smile', votes: '1'},
          {type: 'meh', votes: '0'},
        ],
      },
    ];
    const Screen = render(
      <Provider store={store}>
        <FlatCards flatStore={flats} navigation={navigation} />
      </Provider>,
    );
    const button = Screen.getByTestId('FlatCardsButton');
    fireEvent.press(button);
    expect(Screen).toMatchSnapshot();
  });

  test('When flats_id === user.user.favorites removeFav is invoked', () => {
    flats = [
      {
        _id: 123,
        photos: 'a',
        pros: ['glass', 'folder-open'],
        cons: ['money', 'copy'],
        reviews: ['Mierda de test'],
        face: [
          {type: 'frown', votes: '7'},
          {type: 'grin', votes: '3'},
          {type: 'smile', votes: '1'},
          {type: 'meh', votes: '0'},
        ],
      },
    ];
    const Screen = render(
      <Provider store={store}>
        <FlatCards flatStore={flats} navigation={navigation} />
      </Provider>,
    );
    const button = Screen.getByTestId('FlatCardsButton');
    fireEvent.press(button);
    expect(Screen).toMatchSnapshot();
  });

  test('Then navigation.navigate is invoked', () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <FlatCards flatStore={flats} navigation={navigation} />
      </Provider>,
    );
    const button = getByTestId('FlatCard');
    fireEvent.press(button);
    expect(navigation.navigate).not.toHaveBeenCalled();
  });
});
