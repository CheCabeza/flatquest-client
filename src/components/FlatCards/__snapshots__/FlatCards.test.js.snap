// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a FlatCards component Should render the FlatCards of page 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="FlatCard"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F5F5",
        "borderRadius": 20,
        "display": "flex",
        "elevation": 5,
        "justifyContent": "center",
        "margin": 10,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "a",
        }
      }
      style={
        Object {
          "borderTopLeftRadius": 20,
          "borderTopRightRadius": 20,
          "height": 220,
          "width": 350,
        }
      }
    />
    <View
      style={
        Object {
          "margin": 10,
        }
      }
    >
      <Text />
      <Text />
      <Text>
         €
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <Text>
           
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
           
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
          m2
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰭍
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F5F5F5",
                "borderRadius": 50,
                "display": "flex",
                "height": 45,
                "justifyContent": "center",
                "position": "relative",
                "width": 45,
              },
              Object {
                "right": -30,
                "top": -85,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="FlatCardsButton"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 25,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F5F5F5",
                "borderRadius": 50,
                "display": "flex",
                "height": 45,
                "justifyContent": "center",
                "position": "relative",
                "width": 45,
              },
              Object {
                "right": -10,
                "top": -260,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 35,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#397ED0",
                "borderRadius": 50,
                "display": "flex",
                "height": 20,
                "justifyContent": "center",
                "left": -5,
                "position": "absolute",
                "top": -5,
                "width": 20,
              }
            }
          >
            <Text>
              7
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Given a FlatCards component When flats_id !== user.user.favorites addFav is invoked 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="FlatCard"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F5F5",
        "borderRadius": 20,
        "display": "flex",
        "elevation": 5,
        "justifyContent": "center",
        "margin": 10,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "a",
        }
      }
      style={
        Object {
          "borderTopLeftRadius": 20,
          "borderTopRightRadius": 20,
          "height": 220,
          "width": 350,
        }
      }
    />
    <View
      style={
        Object {
          "margin": 10,
        }
      }
    >
      <Text />
      <Text />
      <Text>
         €
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <Text>
           
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
           
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
          m2
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰭍
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F5F5F5",
                "borderRadius": 50,
                "display": "flex",
                "height": 45,
                "justifyContent": "center",
                "position": "relative",
                "width": 45,
              },
              Object {
                "right": -30,
                "top": -85,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="FlatCardsButton"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 25,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F5F5F5",
                "borderRadius": 50,
                "display": "flex",
                "height": 45,
                "justifyContent": "center",
                "position": "relative",
                "width": 45,
              },
              Object {
                "right": -10,
                "top": -260,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 35,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#397ED0",
                "borderRadius": 50,
                "display": "flex",
                "height": 20,
                "justifyContent": "center",
                "left": -5,
                "position": "absolute",
                "top": -5,
                "width": 20,
              }
            }
          >
            <Text>
              7
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Given a FlatCards component When flats_id === user.user.favorites removeFav is invoked 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="FlatCard"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F5F5",
        "borderRadius": 20,
        "display": "flex",
        "elevation": 5,
        "justifyContent": "center",
        "margin": 10,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "a",
        }
      }
      style={
        Object {
          "borderTopLeftRadius": 20,
          "borderTopRightRadius": 20,
          "height": 220,
          "width": 350,
        }
      }
    />
    <View
      style={
        Object {
          "margin": 10,
        }
      }
    >
      <Text />
      <Text />
      <Text>
         €
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <Text>
           
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
           
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text>
          m2
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 25,
              },
              Object {
                "margin": 10,
              },
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰭍
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F5F5F5",
                "borderRadius": 50,
                "display": "flex",
                "height": 45,
                "justifyContent": "center",
                "position": "relative",
                "width": 45,
              },
              Object {
                "right": -30,
                "top": -85,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="FlatCardsButton"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#EE3BC5",
                    "fontSize": 25,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F5F5F5",
                "borderRadius": 50,
                "display": "flex",
                "height": 45,
                "justifyContent": "center",
                "position": "relative",
                "width": 45,
              },
              Object {
                "right": -10,
                "top": -260,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 35,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5Free-Regular",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "400",
                },
              ]
            }
          >
            
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#397ED0",
                "borderRadius": 50,
                "display": "flex",
                "height": 20,
                "justifyContent": "center",
                "left": -5,
                "position": "absolute",
                "top": -5,
                "width": 20,
              }
            }
          >
            <Text>
              7
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
